### ISSUE TITLE
Minimal runtime support

### ISSUE THREAD

**Issue by tshort** [MikeInnes/Charlotte.jl#13]
Julia's runtime provides a lot of functionality used by base Julia code. A subset of this is needed to support basic data structures like arrays, dicts, and strings. 

Here's a rather intimidating list of `ccall`s from base Julia:

https://gist.github.com/tshort/8d556559a21bf38da44c0c01d0cf9ea3

Hopefully, we can get a lot of functionality by intercepting and replacing a small subset of these.

**Comment by MikeInnes** [comment]
Yeah, this is a useful reference, but I don't think it's nearly as bad as it looks. Most of it is runtime support (e.g. GC and IO) that we don't want from C anyway as WASM will provide it (possibly in future) – and we can get a very long way before it's truly necessary.

Aside from a bunch of other systems and library stuff (bignums, process signals etc) that doesn't make sense for us, Julia does special-case a couple of built in types (arrays and dicts). I'm imagining that we'll get arrays fairly straightforwardly by implementing pointer loads/stores and building an array type on top of that (in Julia) and compiling it. (Pretty much like CUDAnative does with device arrays.) For now we can assume they will be passed as inputs, which avoids any messy memory stuff.

The most ambitious thing is probably having an implementation of Julia's type system and method dispatch. We'll need that eventually, but only once we're happily compiling fully-static Julia code.

**Comment by tshort** [comment]
I agree on the approach of implementing our own array type in Julia. Eventually, I think we'll want to support code that uses Julia arrays, so we'll either have to swap them out for our array type or support more of the libjulia code.

**Comment by MikeInnes** [comment]
> swap them out for our array type

Lucky we have a compiler that will make this very easy :)
